build(custom): fix include path of lv_conf.cmake

AndreCostaaa opened the pull request:
Fixes #8046 

The reason the CI wasn't failing is because the `CMAKE_CURRENT_BINARY_DIR` and `CMAKE_BINARY_DIR` point to the same path when building lvgl alone : 

```bash
#inside lvgl
mkdir build; cmake ..
```

But in other repositories like the `lv_port_pc_vscode`, these paths differ as `CMAKE_CURRENT_BINARY_DIR` will be `lv_port_pc_vscode/build/lvgl` whereas `CMAKE_BINARY_DIR` will be `lv_port_pc_vscode/build` when doing something like

```bash
#inside lv_port_pc_vscode
mkdir build; cmake ..
```

cc @Daxidz

Daxidz commented:
Nice catch, I guess I was to focused on my testing setup. Will be more careful in the future

AndreCostaaa commented:
No worries. Happens to the best of us :)

kisvegabor reviewed (approved):
Thank you!

AndreCostaaa commented:
There were actually two separate issues involved:

Issue 1: As described in the PR, the build would fail when the LVGL CMakeLists.txt wasn’t the first one invoked.
- Fixed by [b6b36a6](https://github.com/lvgl/lvgl/pull/8052/commits/b6b36a6b136b3c40ea81da2160f8c2a1317b2d71)

Issue 2: As reported by @pfvogel in [issue #8046](https://github.com/lvgl/lvgl/issues/8046), the build would break when building LVGL directly inside the lvgl folder with a lv_conf.h also located there. See [this comment](https://github.com/lvgl/lvgl/issues/8046#issuecomment-2778843737) for details.
- Fixed by [5bc874c](https://github.com/lvgl/lvgl/pull/8052/commits/5bc874ca2d4a801f1d0c2688a27166c5036b1a5d)

The third commit also resolves an issue where multiple definitions couldn’t be passed to `lv_conf_internal_gen.py`, which caused the CI to fail due to incorrect handling of LV_CONF_PATH for the tests

kisvegabor reviewed (approved):
Thank you!

It's also related: https://github.com/lvgl/lvgl/pull/8069

Pavi92 commented:
there is also a problem with the generate_cmake_variables.py.
if the line in lv_conf_expanded.h is something like ```#define LV_USE_NAMEX (LV_USE_NAME1 | LV_USE_NAME2 | LV_USE_NAME3)```
the corresponding line in lv_conf.cmake is not generated correctly. It will look like ```set(CONFIG_LV_USE_NAMEX (LV_USE_NAME1)``` and cmake will complain about the missing closing bracket.

```diff
diff --git a/scripts/generate_cmake_variables.py b/scripts/generate_cmake_variables.py
index 33a230c86..cf83f1b4b 100755
--- a/scripts/generate_cmake_variables.py
+++ b/scripts/generate_cmake_variables.py
@@ -92,14 +92,14 @@ def generate_cmake_variables(path_input: str, path_output: str, kconfig: bool):
         # For the rest of config, simply add CONFIG_ and write
         # all LV_USE_* configs, as these are the one needed in cmake
         elif line.startswith(f'{CONFIG_PATTERN}'):
-
+            
             parts = line.split()
             if len(parts) < 3:
                 continue
 
+            value = ' '.join(parts[2:])
+            value = value.strip()
             name = parts[1]
-            value = parts[2].strip()
-
             fout.write(f'set({CONFIG_PREFIX}{name} {value})\n')
 
 

```

AndreCostaaa commented:
I think that your fix won't work because CMake's OR operator is the keyword `OR` and not `|` afaik. We could detect that the line being parsed contains `|` and replace it with `OR` but that will start to get really complicated.

The C language is too rich and I don't think trying to support every possible way of defining a macro needs to be supported in the `generate_cmake_variables.py` script when we can always simplify the `lv_conf.h` defines into `#define LV_USE_XX 1` in one way or another. What do you think?

Pavi92 commented:
I agree it would get really complicated.
Maybe it would be better to to create a file lv_conf.cmake with the default configuration as cache variables and use cmakes configure_file command to create the lv_conf.h from these variables ?

AndreCostaaa commented:
The intended approach is that `lv_conf.h` and `.config` serve as the primary entry points for configuring LVGL—whether manually or through tools like `menuconfig`. Introducing a new `lv_conf.cmake` file as an entrypoint doesn't feel like simplifying the process; in fact, i would argue it complicates it. I could be wrong though

kisvegabor commented:
I merged it to fix the CMake issue

liamHowatt commented:
Thank you :pray:

FYI I opened a PR for something else. I'm open to another solution. I added reproduce info. #8079
